src/mul.cpp

```diff
@@ -325,10 +329,18 @@ RCP<const Basic> mul(const RCP<const Basic> &a, const RCP<const Basic> &b)
     } else {
         RCP<const Basic> exp;
         RCP<const Basic> t;
-        Mul::as_base_exp(a, outArg(exp), outArg(t));
-        Mul::dict_add_term_new(outArg(coef), d, exp, t);
-        Mul::as_base_exp(b, outArg(exp), outArg(t));
-        Mul::dict_add_term_new(outArg(coef), d, exp, t);
+        if (is_a_Number(*a)) {
+            imulnum(outArg(coef), rcp_static_cast<const Number>(a));
+        } else {
+            Mul::as_base_exp(a, outArg(exp), outArg(t));
+            Mul::dict_add_term_new(outArg(coef), d, exp, t);
+        }
+        if (is_a_Number(*b)) {
+            imulnum(outArg(coef), rcp_static_cast<const Number>(b));
+        } else {
+            Mul::as_base_exp(b, outArg(exp), outArg(t));
+            Mul::dict_add_term_new(outArg(coef), d, exp, t);
+        }
```

Comment:
There's a slight difference on what they do. When multiplying `2**x` by `2` previously we got `2**(x+1)`, but multiplying `2**x*y` by `2` we got `(1/2)*2**x*y`. This change makes them all behave as the latter. Therefore avoids the dictionary check.